src/Lykke.Service.PayInvoicePortal/Controllers/Api/SignRequestController.cs

```diff
@@ -0,0 +1,93 @@
+﻿using System;
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading.Tasks;
+using Common;
+using Common.Log;
+using Lykke.Common.Api.Contract.Responses;
+using Lykke.Common.Log;
+using Lykke.Service.PayAuth.Client;
+using Lykke.Service.PayAuth.Client.Models;
+using Lykke.Service.PayInternal.Client;
+using Lykke.Service.PayInternal.Client.Exceptions;
+using Lykke.Service.PayInvoicePortal.Core.Extensions;
+using Lykke.Service.PayInvoicePortal.Models.SignRequest;
+using Lykke.Service.PayMerchant.Client;
+using LykkePay.Common.Validation;
+using Microsoft.AspNetCore.Mvc;
+
+namespace Lykke.Service.PayInvoicePortal.Controllers.Api
+{
+    [Route("api/signrequest")]
+    public class SignRequestController : Controller
+    {
+        public const string SystemId = "LykkePay";
+        private readonly IPayAuthClient _payAuthClient;
+        private readonly IPayMerchantClient _payMerchantClient;
+        private readonly ILog _log;
+
+        public SignRequestController(
+            IPayAuthClient payAuthClient,
+            IPayMerchantClient payMerchantClient,
+            ILogFactory logFactory)
+        {
+            _payAuthClient = payAuthClient;
+            _payMerchantClient = payMerchantClient;
+            _log = logFactory.CreateLog(this);
+        }
+
+        [HttpPost]
+        [ValidateModel]
+        public async Task<IActionResult> SignRequest([FromBody] SignRequestModel model)
+        {
+            try
+            {
+                var merchant = await _payMerchantClient.Api.GetByIdAsync(model.LykkeMerchantId);
+
+                if (model.ApiKey != merchant.ApiKey)
+                    throw new InvalidOperationException("Invalid api key for merchant.");
+
+                var rsa = model.RsaPrivateKey.CreateRsa();
+
+                var signedBody = Convert.ToBase64String(
+                    rsa.SignData(
+                        Encoding.UTF8.GetBytes($"{model.ApiKey}{model.Body}"),
+                        HashAlgorithmName.SHA256,
+                        RSASignaturePadding.Pkcs1));
+
+                // check signed body
+                var verifyRequest = new VerifyRequest
+                {
+                    ClientId = model.LykkeMerchantId,
+                    SystemId = SystemId,
+                    Signature = signedBody,
+                    Text = model.Body
+                };
+
+                var verificationResponse = await _payAuthClient.VerifyAsync(verifyRequest);
+
+                if (!verificationResponse.Description.Equals("OK", StringComparison.InvariantCultureIgnoreCase))
```

Comment:
Wouldn't it be better to check SecurityErrorType instead of strings comparison?

Reply:
For now OK for me is enough. In the future will use if the necessity appears to check and handle other cases.